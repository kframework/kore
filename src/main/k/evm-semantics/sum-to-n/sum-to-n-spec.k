requires "asm.k"
requires "edsl.k"

module VERIFICATION
    imports EDSL
    imports EVM-ASSEMBLY

    rule #sizeWordStack ( WS , N:Int )
      => N +Int #sizeWordStack ( WS , 0 )
      requires N =/=K 0
      [lemma]

    syntax ByteArray ::= "sumToN" [function]
 // ----------------------------------------
    rule sumToN
      => #asmOpCodes(PUSH(1, 0) ; SWAP(1)                   // s = 0 ; n = N
                    ; JUMPDEST                              // label:loop
                    ; DUP(1) ; ISZERO ; PUSH(1, 20) ; JUMPI // if n == 0, jump to end
                    ; DUP(1) ; SWAP(2) ; ADD                // s = s + n
                    ; SWAP(1) ; PUSH(1, 1) ; SWAP(1) ; SUB  // n = n - 1
                    ; PUSH(1, 3) ; JUMP                     // jump to loop
                    ; JUMPDEST                              // label:end
                    ; .OpCodes
                    ) [macro]
endmodule

module SUM-TO-N-SPEC
    imports VERIFICATION

    rule <k> #execute ... </k>
         <mode> NORMAL </mode>
         <schedule> DEFAULT </schedule>
         <callStack> .List </callStack>
         <memoryUsed> 0   </memoryUsed>
         <localMem> .Map </localMem>
         <callGas> _ => _ </callGas>
         <program> sumToN </program>
         <jumpDests> JUMPDESTS </jumpDests>

         <pc>        0  => 21                                </pc>
         <wordStack> N : WS => 0 : N *Int (N +Int 1) /Int 2 : WS </wordStack>
         <gas>       G  => G -Int (52 *Int N +Int 27)        </gas>
         <static> STATIC </static>

      requires N >=Int 0
       andBool N <=Int 340282366920938463463374607431768211455
       andBool 0 <=Int #sizeWordStack(WS)
       andBool #sizeWordStack(WS) <Int 1021
       andBool G >=Int 52 *Int N +Int 27
       andBool JUMPDESTS ==K #computeValidJumpDests(sumToN)

endmodule
