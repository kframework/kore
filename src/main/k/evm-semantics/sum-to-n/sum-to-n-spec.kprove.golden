  <kevm>
    <k>
      #next [ PUSH ( 1 ) ] ~> #execute ~> DotVar2
    </k>
    DotVar1
    <mode>
      NORMAL
    </mode>
    <schedule>
      DEFAULT
    </schedule>
    <ethereum>
      <evm>
        _7
        _8
        <callStack>
          .List
        </callStack>
        _9
        _10
        <callState>
          <program>
            96 : 0 : 144 : 91 : 128 : 21 : 96 : 20 : 87 : 128 : 145 : 1 : 144 : 96 : 1 : 144 : 3 : 96 : 3 : 86 : 91 : .WordStack
          </program>
          <jumpDests>
            SetItem ( 3 )
            SetItem ( 20 )
          </jumpDests>
          _2
          _3
          _4
          _5
          <wordStack>
            N : WS
          </wordStack>
          <localMem>
            .Map
          </localMem>
          <pc>
            0
          </pc>
          <gas>
            G
          </gas>
          <memoryUsed>
            0
          </memoryUsed>
          <callGas>
            _0
          </callGas>
          <static>
            STATIC
          </static>
          _6
        </callState>
        _11
        _12
        _13
        _14
        _15
      </evm>
      DotVar3
    </ethereum>
  </kevm>
#And
  {
    JUMPDESTS
  #Equals
    SetItem ( 3 )
    SetItem ( 20 )
  }
#And
  {
    true
  #Equals
    N >=Int 0 andBool N <=Int 340282366920938463463374607431768211455 andBool 0 <=Int #sizeWordStack ( WS , 0 ) andBool #sizeWordStack ( WS , 0 ) <Int 1021 andBool G >=Int 52 *Int N +Int 27
  }
